core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -0,0 +1,399 @@
+///
+/// Each I/O threads will have its own copy of `TriePrefetchingStorage`, so
+/// this should remain a cheap object.
+#[derive(Clone)]
+struct TriePrefetchingStorage {
+    /// Store is shared with parent `TrieCachingStorage`.
+    store: Store,
+    shard_uid: ShardUId,
+    /// Shard cache is shared with parent `TrieCachingStorage`. But the
+    /// pre-fetcher uses this in read-only mode to avoid premature evictions.
+    shard_cache: TrieCache,
+    /// Shared with parent `TrieCachingStorage`.
+    prefetching: PrefetchStagingArea,
+}
+
+/// This type is shared between runtime crate and store crate.
+///
+/// The former puts requests in, the latter serves requests.
+/// With this API, the store does not know about receipts etc, and the runtime
+/// does not know about the trie structure. The only thing they share is this object.
+#[derive(Clone)]
+pub struct PrefetchApi {
+    /// Bounded, shared queue for all IO threads to take work from.
+    ///
+    /// Work items are defined as `TrieKey` because currently the only
+    /// work is to prefetch a trie key. If other IO work is added, consider
+    /// changing the queue to an enum.
+    /// The state root is also included because multiple chunks could be applied
+    /// at the same time.
+    work_queue_tx: crossbeam::channel::Sender<(StateRoot, TrieKey)>,
+    work_queue_rx: crossbeam::channel::Receiver<(StateRoot, TrieKey)>,
+    /// Prefetching IO threads will insert fetched data here. This is also used
+    /// to mark what is already being fetched, to avoid fetching the same data
+    /// multiple times.
+    pub(crate) prefetching: PrefetchStagingArea,
+}
+
+/// Staging area for in-flight prefetch requests and a buffer for prefetched data.
+///
+/// Before starting a pre-fetch, a slot is reserved for it. Once the data is
+/// here, it will be put in that slot. The parent `TrieCachingStorage` needs
+/// to take it out and move it to the shard cache.
+///
+/// A shared staging area is the interface between `TrieCachingStorage` and
+/// `TriePrefetchingStorage`. The parent simply checks the staging area before
+/// going to the DB. Otherwise, no communication between the two is necessary.
+///
+/// This design also ensures the shard cache works exactly the same with or
+/// without the prefetcher, because the order in which it sees accesses is
+/// independent of the prefetcher.
+#[derive(Default, Clone)]
+pub(crate) struct PrefetchStagingArea(Arc<Mutex<InnerPrefetchStagingArea>>);
+
+#[derive(Default)]
+struct InnerPrefetchStagingArea {
+    slots: HashMap<CryptoHash, PrefetchSlot>,
+    size_bytes: usize,
+}
+
+/// Result when atomically accessing the prefetch staging area.
+pub(crate) enum PrefetcherResult {
+    SlotReserved,
+    Pending,
+    Prefetched(Arc<[u8]>),
+    MemoryLimitReached,
+}
+
+/// Type used interanlly in the staging area to keep track of requests.
+#[derive(Clone, Debug)]
+enum PrefetchSlot {
+    PendingPrefetch,
+    PendingFetch,
```

Comment:
Prefetching is always done by hash, and hashes must be present in node storage. So if there is a failure, it is an actual IO error or a missing node, neither of which we can handle anyway.
I've now changed it that we will at least remove the reserved slot in case of a failure, but I don't see any value in putting the error value in there. Does that make sense?